diff --git c/modules/tools/debugger/config.el w/modules/tools/debugger/config.el
index 7483b54d7..4131773d1 100644
--- c/modules/tools/debugger/config.el
+++ w/modules/tools/debugger/config.el
@@ -12,7 +12,7 @@
     ((:lang rust +lsp)       :after rustic      :require (dap-lldb dap-cpptools))
     ((:lang javascript +lsp)
      :after (js2-mode typescript-mode)
-     :require (dap-node dap-chrome dap-firefox ,@(if (featurep :system 'windows) '(dap-edge)))))
+     :require (dap-js dap-firefox)))
   "An alist mapping Doom modules to major mode and DAP packages.")
 
 
